src/Microsoft.AspNet.Mvc.TagHelpers/Internal/GlobbingUrlBuilder.cs

```diff
@@ -73,93 +80,104 @@ public GlobbingUrlBuilder(IFileProvider fileProvider, IMemoryCache cache, PathSt
         /// <param name="includePattern">The file globbing include pattern.</param>
         /// <param name="excludePattern">The file globbing exclude pattern.</param>
         /// <returns>The list of URLs</returns>
-        public virtual ICollection<string> BuildUrlList(string staticUrl, string includePattern, string excludePattern)
+        public virtual IReadOnlyList<string> BuildUrlList(
+            string staticUrl,
+            string includePattern,
+            string excludePattern)
         {
-            var urls = new HashSet<string>(StringComparer.Ordinal);
+            // Get urls that match the globbing patterns specified
+            var globbedUrls = ExpandGlobbedUrl(includePattern, excludePattern);
 
-            // Add the statically declared url if present
-            if (staticUrl != null)
+            if (staticUrl == null)
             {
-                urls.Add(staticUrl);
+                return globbedUrls;
             }
 
-            // Add urls that match the globbing patterns specified
-            var matchedUrls = ExpandGlobbedUrl(includePattern, excludePattern);
-            urls.UnionWith(matchedUrls);
+            // The staticUrl always appears first in the sequence.
+            var urls = new List<string>(1 + globbedUrls.Count)
+            {
+                staticUrl
+            };
```

Comment:
Suggest changing `ExpandGlobbedUrl()` to return a concrete `List<string>` and `Insert(0, staticUrl)` by default. Allocate a new `List<string>` only if `ReferenceEquals(EmptyList, globbedUrls)`. Of course `EmptyList` would need to be a `List<string>` then.
Note I'm not suggesting you change the `BuildUrlList()` return type.